File: 프로그래머스/2/60057. 문자열 압축/문자열 압축.cpp
#include <bits/stdc++.h>
using namespace std;

int solution(string s) 
{
    int ans = 1e9;
    
    for (int len = 1; len <= s.size(); len++)
    {
        // if (s.size() % len) continue;
        int idx = 0, cnt = 0, val = 0;
        string res = "";
        
        while (true)
        {
            if (idx + len - 1 >= s.size())
            {
                val += s.substr(idx).size();
                break;
            }
            
            string tmp = s.substr(idx, len);
            // cout << res << ' ' << tmp << endl;
            if (res == tmp) cnt++;
            else 
            {
                if (cnt > 1)
                    val += to_string(cnt).size();
                val += res.size();
                res = tmp; cnt = 1;
            }
            idx += len;
        }
        
        // cout << cnt << ' ' << len << endl;
        
        if (cnt > 1)
            val += to_string(cnt).size();
        
        val += res.size();
        ans = min(ans, val);
    }

    return ans; 
}

File: 프로그래머스/2/60057. 문자열 압축/문자열 압축.java
import java.util.*;

class Solution 
{
    public int solution(String s) 
    {
        int ans = (int)1e9;
        for (int len = 1; len <= s.length(); len++)
        {
            int res = 0;
            String prv = s.substring(0, len);
            int cnt = 1;
            for (int i = len; i < s.length(); i += len)
            {
                String tmp = s.substring(i, Math.min(i + len, s.length()));
                if (prv.equals(tmp))
                    cnt++;
                else
                {
                    res += len;
                    if (cnt > 1)
                        res += String.valueOf(cnt).length();
                    cnt = 1;
                }
                prv = tmp;
            }
            
            res += prv.length();
            if (cnt > 1)
                res += String.valueOf(cnt).length();
            ans = Math.min(ans, res);
        }
        
        return ans;
    }
}